In [485]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [486]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [487]:
dataset = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/matches.csv - Sheet1.csv')

In [488]:
dataset.shape

(4037, 9)

In [489]:
dataset.isnull().sum()

team1               0
team2               0
innings1_overs      0
innings1_wickets    0
innings1_runs       0
innings2_overs      0
innings2_wickets    0
innings2_runs       0
year                0
dtype: int64

In [490]:
dataset.insert(9, "result", np.nan)
dataset.head()

team1         team2  innings1_overs  ...  innings2_runs  year  result
0    SRI LANKA      ZIMBABWE            50.0  ...            213  2001     NaN
1  NETHERLANDS        CANADA            50.0  ...            172  2007     NaN
2     ZIMBABWE  SOUTH AFRICA            50.0  ...            175  2003     NaN
3        INDIA   NEW ZEALAND            49.0  ...            236  2010     NaN
4    AUSTRALIA   NEW ZEALAND            48.4  ...            182  2009     NaN

[5 rows x 10 columns]

In [491]:
for i in range(0,4037):
  if dataset.innings1_runs[i] > dataset.innings2_runs[i]:
    dataset.result[i] = "won"
  else:
    dataset.result[i] = "lost"

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-vie

In [492]:
dataset.head()

team1         team2  innings1_overs  ...  innings2_runs  year  result
0    SRI LANKA      ZIMBABWE            50.0  ...            213  2001     won
1  NETHERLANDS        CANADA            50.0  ...            172  2007     won
2     ZIMBABWE  SOUTH AFRICA            50.0  ...            175  2003    lost
3        INDIA   NEW ZEALAND            49.0  ...            236  2010     won
4    AUSTRALIA   NEW ZEALAND            48.4  ...            182  2009    lost

[5 rows x 10 columns]

In [493]:
dataset.tail()

team1         team2  innings1_overs  ...  innings2_runs  year  result
4032      ENGLAND     AUSTRALIA            50.0  ...            221  2013     won
4033    AUSTRALIA         INDIA            50.0  ...            230  2009    lost
4034  NEW ZEALAND      ZIMBABWE            48.3  ...            158  2012     won
4035    AUSTRALIA  SOUTH AFRICA            47.3  ...            218  1996    lost
4036        KENYA      SCOTLAND            50.0  ...            253  2007     won

[5 rows x 10 columns]

In [494]:
from sklearn import preprocessing 
label_encoder = preprocessing.LabelEncoder() 
dataset['team1']= label_encoder.fit_transform(dataset['team1'])
dataset['team2']= label_encoder.fit_transform(dataset['team2'])
dataset['result'] = label_encoder.fit_transform(dataset['result'])

In [495]:
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [496]:
print(X)

[[2.200e+01 2.700e+01 5.000e+01 ... 1.000e+01 2.130e+02 2.001e+03]
 [1.600e+01 6.000e+00 5.000e+01 ... 1.000e+01 1.720e+02 2.007e+03]
 [2.600e+01 2.200e+01 5.000e+01 ... 1.000e+00 1.750e+02 2.003e+03]
 ...
 [1.700e+01 2.700e+01 4.830e+01 ... 1.000e+01 1.580e+02 2.012e+03]
 [3.000e+00 2.200e+01 4.730e+01 ... 8.000e+00 2.180e+02 1.996e+03]
 [1.300e+01 2.100e+01 5.000e+01 ... 8.000e+00 2.530e+02 2.007e+03]]


In [497]:
print(y)

[1 1 0 ... 1 0 1]


In [498]:
dataset.head()

team1  team2  innings1_overs  ...  innings2_runs  year  result
0     22     27            50.0  ...            213  2001       1
1     16      6            50.0  ...            172  2007       1
2     26     22            50.0  ...            175  2003       0
3     11     17            49.0  ...            236  2010       1
4      3     17            48.4  ...            182  2009       0

[5 rows x 10 columns]

In [499]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 1)
print(X_train)
print(X_test)
print(y_train)
print(y_test)

[[  18.    26.    45.  ...    7.   191.  1986. ]
 [  21.    19.    35.  ...   10.   121.  2000. ]
 [   3.    17.    50.  ...   10.   206.  2006. ]
 ...
 [  13.    16.    50.  ...    3.   248.  2009. ]
 [   6.    21.    44.3 ...    8.   209.  2007. ]
 [  21.    26.    42.  ...   10.   230.  2015. ]]
[[  21.     3.    50.  ...    7.   264.  2009. ]
 [   8.    19.    55.  ...   10.    85.  1978. ]
 [  22.    26.    48.3 ...    4.   209.  2013. ]
 ...
 [  11.    23.    50.  ...    8.   265.  2014. ]
 [  18.    17.    50.  ...   10.   227.  2011. ]
 [   3.    11.    50.  ...    5.   261.  2011. ]]
[1 1 1 ... 0 0 1]
[1 1 0 0 0 0 0 1 1 0 0 1 0 1 1 1 0 1 1 0 1 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 1
 0 1 1 0 0 0 0 0 0 0 1 0 1 0 1 0 1 0 1 1 0 1 0 1 1 1 0 1 1 1 0 1 1 1 0 0 1
 1 0 1 1 1 0 0 0 0 1 0 1 0 1 1 0 1 0 1 0 1 0 1 1 1 0 1 1 0 1 0 0 0 1 1 1 1
 0 0 1 1 0 1 1 0 1 0 0 0 1 0 0 1 0 1 0 1 1 1 1 0 1 0 1 0 1 0 1 0 0 0 1 0 0
 1 0 1 1 0 0 0 0 0 1 1 1 1 1 1 0 1 0 0 0 0 0 0 0 0 1 0 1 1 0 1 0 1 0 1 1 0
 0 0 0 

In [500]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train[:,:] = sc.fit_transform(X_train[:,:])
X_test[:,:] = sc.transform(X_test[:,:])
print(X_train)
print(X_test)

[[ 0.39825362  1.26796889 -0.49510918 ...  0.10774435 -0.12923812
  -1.54844699]
 [ 0.78388765  0.39985534 -2.29968245 ...  1.05744359 -1.35435522
  -0.21385987]
 [-1.52991653  0.15182289  0.40717745 ...  1.05744359  0.13328697
   0.35810604]
 ...
 [-0.24446976  0.02780667  0.40717745 ... -1.15852132  0.86835723
   0.644089  ]
 [-1.1442825   0.64788778 -0.62142931 ...  0.42431076  0.18579199
   0.4534337 ]
 [ 0.78388765  1.26796889 -1.03648116 ...  1.05744359  0.55332712
   1.21605491]]
[[ 0.78388765 -1.58440421  0.40717745 ...  0.10774435  1.148384
   0.644089  ]
 [-0.88719315  0.39985534  1.30946408 ...  1.05744359 -1.98441544
  -2.31106821]
 [ 0.91243233  1.26796889  0.1004     ... -0.8419549   0.18579199
   1.02539961]
 ...
 [-0.50155912  0.89592023  0.40717745 ...  0.42431076  1.16588567
   1.12072726]
 [ 0.39825362  0.15182289  0.40717745 ...  1.05744359  0.5008221
   0.8347443 ]
 [-1.52991653 -0.59227444  0.40717745 ... -0.52538849  1.09587898
   0.8347443 ]]


## Naive bayes


In [501]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [502]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [1 0]
 [1 1]
 [1 0]]


In [503]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[337  71]
 [ 37 363]]


0.8663366336633663

In [504]:
predictions_NB = classifier.predict(X_test)
from sklearn.metrics import mean_absolute_error, accuracy_score, f1_score, recall_score, precision_score
mae = mean_absolute_error(y_test, predictions_NB)
acc = accuracy_score(y_test, predictions_NB)
f1_score = f1_score(y_test, predictions_NB,average='weighted')
recall = recall_score(y_test, predictions_NB,average='weighted')
precision = precision_score(y_test, predictions_NB,average='weighted')
print("Mean Absolute Error : ",mae)
print("Accuracy: ",acc)
print("Recall: ",recall)
print("Precision: ",precision)
print("F1-score: ",f1_score)

Mean Absolute Error :  0.13366336633663367
Accuracy:  0.8663366336633663
Recall:  0.8663366336633663
Precision:  0.8690576430914981
F1-score:  0.8661554613390647


## ExtraTrees Classifier

In [505]:
from sklearn.ensemble import ExtraTreesClassifier
model_ET = ExtraTreesClassifier(max_features='log2',min_samples_split=3)
model_ET.fit(X_train, y_train)

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=None, max_features='log2',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_impurity_split=None,
                     min_samples_leaf=1, min_samples_split=3,
                     min_weight_fraction_leaf=0.0, n_estimators=100,
                     n_jobs=None, oob_score=False, random_state=None, verbose=0,
                     warm_start=False)

In [506]:
y_pred = model_ET.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [1 1]
 [0 0]]


In [507]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[392  16]
 [ 31 369]]


0.9418316831683168

In [508]:
predictions_ET = model_ET.predict(X_test)
from sklearn.metrics import mean_absolute_error, accuracy_score, f1_score, recall_score, precision_score
mae = mean_absolute_error(y_test, predictions_ET)
acc = accuracy_score(y_test, predictions_ET)
f1_score = f1_score(y_test, predictions_ET,average='weighted')
recall = recall_score(y_test, predictions_ET,average='weighted')
precision = precision_score(y_test, predictions_ET,average='weighted')
print("Mean Absolute Error : ",mae)
print("Accuracy: ",acc)
print("Recall: ",recall)
print("Precision: ",precision)
print("F1-score: ",f1_score)

Mean Absolute Error :  0.05816831683168317
Accuracy:  0.9418316831683168
Recall:  0.9418316831683168
Precision:  0.9424206858375807
F1-score:  0.9418009197272827


## Random Forest

In [509]:
from sklearn.ensemble import RandomForestClassifier
model_RF = RandomForestClassifier(n_estimators = 100, n_jobs=-1)
# Add silent=True to avoid printing out updates with each cycle
model_RF.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=-1, oob_score=False, random_state=None, verbose=0,
                       warm_start=False)

In [510]:
y_pred = model_RF.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [1 1]
 [0 0]]


In [511]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[396  12]
 [ 19 381]]


0.9616336633663366

## Random Forest using GridSearchCV 

In [530]:

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV,StratifiedKFold

model_rc = RandomForestClassifier()

parameters = {'n_estimators': [1000,1100,1200]       
              }


clf = GridSearchCV(model_rc, parameters, n_jobs = 4,
                   cv=StratifiedKFold(n_splits=2), 
                   scoring='accuracy',
                   verbose=2, refit=True)

clf.fit(X,y)
print(clf.best_score_)
print(clf.best_params_)

Fitting 2 folds for each of 3 candidates, totalling 6 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done   3 out of   6 | elapsed:   11.1s remaining:   11.1s
[Parallel(n_jobs=4)]: Done   6 out of   6 | elapsed:   17.1s finished


0.9524402958809054
{'n_estimators': 1200}


In [531]:
print(clf.best_score_)
print(clf.best_params_)

0.9524402958809054
{'n_estimators': 1200}


## Training the Random Forest Model again using the best Parameters found from Grid search CV

In [535]:
from sklearn.ensemble import RandomForestClassifier

model_RF = RandomForestClassifier(n_estimators =1200)
# Add silent=True to avoid printing out updates with each cycle
model_RF.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1200,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [536]:
y_pred = model_RF.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [1 1]
 [0 0]]


In [537]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[397  11]
 [ 19 381]]


0.9628712871287128

In [538]:
predictions_RF = model_RF.predict(X_test)
from sklearn.metrics import mean_absolute_error, accuracy_score, f1_score, recall_score, precision_score
mae = mean_absolute_error(y_test, predictions_RF)
acc = accuracy_score(y_test, predictions_RF)
f1_score = f1_score(y_test, predictions_RF,average='weighted')
recall = recall_score(y_test, predictions_RF,average='weighted')
precision = precision_score(y_test, predictions_RF,average='weighted')
print("Mean Absolute Error : ",mae)
print("Accuracy: ",acc)
print("Recall: ",recall)
print("Precision: ",precision)
print("F1-score: ",f1_score)

Mean Absolute Error :  0.03712871287128713
Accuracy:  0.9628712871287128
Recall:  0.9628712871287128
Precision:  0.96304566190528
F1-score:  0.9628640048470867


## XG Boost Classifier

In [517]:
from xgboost import XGBClassifier
model_XG = XGBClassifier()
model_XG.fit(X_train, y_train, verbose=2)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [518]:
y_pred = model_XG.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [1 1]
 [0 0]]


In [519]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[399   9]
 [ 16 384]]


0.969059405940594

## XGBoost using GridSearchCV

In [520]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV,StratifiedKFold

xgb_model = XGBClassifier()


parameters = {'learning_rate': [0.1], #so called `eta` value
              'n_estimators': [650,700,750],  #number of trees, change it to 1000 for better results
              'max_depth': [3,4,5]
              }


clf = GridSearchCV(xgb_model, parameters, n_jobs = 6,
                   cv=StratifiedKFold(n_splits=2), 
                   scoring='accuracy',
                   verbose=2, refit=True)

clf.fit(X,y)

Fitting 2 folds for each of 9 candidates, totalling 18 fits


[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.
[Parallel(n_jobs=6)]: Done  18 out of  18 | elapsed:   14.3s finished


GridSearchCV(cv=StratifiedKFold(n_splits=2, random_state=None, shuffle=False),
             error_score=nan,
             estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                     colsample_bylevel=1, colsample_bynode=1,
                                     colsample_bytree=1, gamma=0,
                                     learning_rate=0.1, max_delta_step=0,
                                     max_depth=3, min_child_weight=1,
                                     missing=None, n_estimators=100, n_jobs=1,
                                     nthread=None, objective='binary:logistic',
                                     random_state=0, reg_alpha=0, reg_lambda=1,
                                     scale_pos_weight=1, seed=None, silent=None,
                                     subsample=1, verbosity=1),
             iid='deprecated', n_jobs=6,
             param_grid={'learning_rate': [0.1], 'max_depth': [3, 4, 5],
                         'n_estima

In [521]:
print(clf.best_score_)
print(clf.best_params_)

0.971513437016333
{'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 700}


## Training the XGBoost model again using the best Parameters found from Grid search CV

In [522]:
from xgboost import XGBClassifier
model_XG = XGBClassifier(learning_rate= 0.1, max_depth= 3, n_estimators= 700)

model_XG.fit(X_train, y_train, verbose=2)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=700, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [523]:
y_pred = model_XG.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [0 0]
 ...
 [0 0]
 [1 1]
 [0 0]]


In [524]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[403   5]
 [  9 391]]


0.9826732673267327

In [525]:
predictions_XG = model_XG.predict(X_test)
from sklearn.metrics import mean_absolute_error, accuracy_score, f1_score, recall_score, precision_score
mae = mean_absolute_error(y_test, predictions_XG)
acc = accuracy_score(y_test, predictions_XG)
f1_score = f1_score(y_test, predictions_XG,average='weighted')
recall = recall_score(y_test, predictions_XG,average='weighted')
precision = precision_score(y_test, predictions_XG,average='weighted')
print("Mean Absolute Error : ",mae)
print("Accuracy: ",acc)
print("Recall: ",recall)
print("Precision: ",precision)
print("F1-score: ",f1_score)

Mean Absolute Error :  0.017326732673267328
Accuracy:  0.9826732673267327
Recall:  0.9826732673267327
Precision:  0.9827189029582569
F1-score:  0.9826719931461217


## After analysis using Naive bayes (accuracy = 86.6% ),ExtraTrees Classifier(accuracy = 94.1% ), RandomForest Classifier (accuracy = 96.2% ) and XGBoost Classifier (accuracy = 98.2% ),it can be inferred that XGBoost is the best model because it provides the best accuracy and has improved the prediction of the current model,Naive Bayes which was used in the Research Paper.